### DISCUSSION TITLE
Custom container components which observe properties on @children

### DISCUSSION THREAD

**Discussion by mitchgrout** [slint-ui/slint#4806] [Q&A]
The [documentation](https://slint.dev/releases/1.1.0/docs/slint/src/concepts/container) is currently a bit light on how `@children` can be used. I'd like to make an image-carousel-like container component, which would require me to do things like...

- Select the maximum width of all children, and
- Determine how many children there are

My gut feeling would be that these properties would be available on `@children`, like `width: max(@children.width...)`, or perhaps it could be attained by iterating with `@children` being a [model](https://slint.dev/releases/1.1.0/docs/slint/src/reference/repetitions), but these all appear to be parse errors. I've noted that the core container components like `HorizontalLayout` seem to be [implemented outside of the Slint language](https://github.com/slint-ui/slint/blob/b7e437fb370a08e5d7ef6b3e1c853438c6557add/tools/lsp/language/component_catalog.rs#L17-L24), so I suspect this might not be doable in v1.4?

If this is currently not doable, is it worthwhile implementing? It would allow for richer components, but at the cost of increased complexity.

**Comment by hunger** [comment]
Right now `@children` is just a "add children *here*" marker for the compiler.

I would expect you having your images in a model, so can't you just look at the data in that model?

**Reply by mitchgrout** [reply]
I could definitely have something like `in property<[string]> images` or similar, but I'm interested in whether a more generic solution is possible or desirable. For example, I might want to display images on the first `N-1` screens, and on the last screen have a text+button component.

**Reply by hunger** [reply]
Screens?

I do not see much value in having a model with arbitrary data in it. A model on `@children` would probably not have too much useful contents... it could contain pretty arbitrary elements after all. Since the element tree in slint is fixed, and many properties get optimized out you can also not do too much with the data in that model.
